Package lib

autra opened the pull request:
Here is the PR that fixes #243 

I've put a lot of explanations on each commit messages. Please do read them!

What I like:
- now our npm package are bundle/compiler agnostics. Client code can use the bundle like an old webpage, or they can require each module separately without relying on our babel config, or even on webpack loaders for glsl (as they are inlined)
- No change from our previous dev workflow
- Minimal addition to our devDependencies, and the ability to put back our real dependencies in "dependencies" tag in `package.json`

What I like less:
- There is now a bit of magic involved (webpack and npm calls babel to process js files, which in turns uses a webpack loader in `webpack-babel.config.js` to inline glsl files)

What still needs to be done:
- Build an itowns-full.js containing our code + all the dependencies, and an itowns.js without dependencies + vendor.js. Then github releases would contain these 3 files, but npm packages might contain only itowns.js (as they would have dependencies through npm)

@tbroyer this is my attempt to realize [agnostic.md](https://gist.github.com/mattdesl/aaf759da84cc44c22305) goals.

@iTowns/reviewers please r?

autra committed:
Chore: Use babel to inline glsl file using webpack loader

As we need to inline glsl files for packaging, but *without* bundling
them like webpack does, we just deal with glsl inlining with babel
instead of webpack when building our lib folder. We need a way to tell
babel where and how to process these glsl though, and that's where
babel-plugin-webpack-loaders comes into play.

We still let webpack deal with glsl files when building our bundles, or
when using webpack-dev-server, hence the custom babel config in
webpack.config.js

tbroyer commented on webpack.config.js:
```diff
@@ -57,7 +49,7 @@ module.exports = {
   },
   resolve: {
-    root: path.resolve(__dirname, 'src'),
+    root: path.resolve(__dirname, ''),
```
Should `resolve.root` be deleted instead? (or set to an empty array) AFAICT, we now want "module names" to only be resolved as Node modules.

tbroyer commented on webpack.config.js:
```diff
@@ -35,14 +35,6 @@ module.exports = {
         loader: 'babel'
       },
-       {
-        test: /\.glsl$/,
```
AIUI, this is no longer need because it's now taken care of by the `babel-plugin-webpack-loaders`, but the README for that project says that it shouldn't be used during a Webpack compilation: https://github.com/istarkov/babel-plugin-webpack-loaders#warning (that plugin runs Webpack synchronously during the Babel transformation, which is itself run by Webpack for bundling)

Maybe it'd be better to have different Babel configurations depending on the usage: when calling babel-cli to produce CommonJS, then use `.babelrc` that calls the babel-plugin-webpack-loader with the babel-webpack.config.js; and when calling webpack, use the raw-loader in webpack.config.js and a different babel config that does not use the babel-plugin-webpack-loader (by inlining the babel config in the babel-loader's `query` and setting `babelrc: false`).

autra replied:
yep, I've seen that but as we are using it for dev and I didn't see any noticeable performance hit, I thought it was not worth the effort. Also I wanted prod and dev build to stay as close as possible.

I did the measurement again, and this time my build is 6sec longer! So I guess the warning is real :-) I'm amending this right now.

tbroyer reviewed (commented):
I really like your work to transform import paths; this should probably be split out into its own commit (along with the `resolve.root` change –see comment below about it though), possibly even its own PR.

autra commented:
> [...] transform import paths; this should probably be split out into its own commit (along with the resolve.root change –see comment below about it though), possibly even its own PR.

I would let this changeset into this Pr, because the reason I'm doing this is solely because of the lib/ stuff. Otherwise, we wouldn't care and would stay with absolute import, as they are easier to deal with internally, right? I'm putting them into their own commit.

autra committed:
Change import path from absolute to relative

This change is a necessary step to allow client code to import our
modules directly.

To achieve this, I used a little python script. I also had to change
some import orders to make the linter happy.

Work on generating lib/ and inlining glsl will be done in a subsequent
commit.

For reference, here is the python script:

import glob
import os
import re
import fileinput

prefix=os.path.join(os.getcwd(), 'src')
list = ('Core', 'Globe', 'MobileMapping', 'Renderer', 'Scene')

for filename in glob.glob(prefix + '/**/*.js', recursive=True):
    for line in fileinput.input(filename, inplace=True):
        # check if local import
        if line.startswith('import ') and any(word in line for word in list):
            search = re.search('import (.*) from \'(.*)\';', line)
            if search:
                # importing a js
                importString = search.group(1)
                import_abs_path = os.path.join(prefix, search.group(2))
                import_rel_path = os.path.relpath(import_abs_path, os.path.dirname(filename))
                if not import_rel_path.startswith('.'):
                    # webpack needs ./ before relative path
                    import_rel_path = './' + import_rel_path
                print('import ' + importString + ' from \'' + import_rel_path + '\';')
            else:
                # importing something else
                print(line, end='')
        else:
            # import from node_modules
            print(line, end='')

autra committed:
Package babelized src/ into lib/

This will allow client code to import our sources directly, without
relying on our bundler or our babel config.

Atm glsl are not inlined, so client code will need some kind of loader
for them. It will be done in a subsequent commit.

autra commented:
@tbroyer I've updated my PR with your required change. Pro: build is now as fast as before. Cons: now, we have 2 conf duplications : babel conf is in `.babelrc` and `webpack.config.js`, glsl loader conf is in `webpack.config.js` and `webpack-babel.config.js`. Do you have an idea to mitigate that?

tbroyer commented:
Well, Webpack configs are JS so they can `require()` other files; this means `webpack.config.js` could possibly `require('./.babelrc')` (no sure how it'll deal with the "JSON5" syntax; worst case would be using `require('json5')` and explicitly parsing the `.babelrc` file) and then _massage_ the data to remove the specific values we don't want to use in Webpack. And similarly `webpack.config.js` could `require('webpack-babel.config.js')` (or an intermediate, shared config) to merge it in. That'd remove duplications but would likely make the configuration more complex. Not sure it'd be worth it (a comment in the files telling you to keep files in sync could possibly work better from a maintenance point of view)

autra committed:
Chore: add comments in webpack configs

autra commented:
@tbroyer I've added some comments. I'm not very comfortable with that, as I always find that nobody reads comments I've ever left in code, but the alternative is meh, so let's try that out.

tbroyer commented on .babelrc:
```diff
@@ -1,4 +1,7 @@
 {
-  "presets": ["es2015"],
-  "plugins": ["transform-runtime"]
+    "presets": ["es2015"],
```
Add a comment here ([this is _JSON5_](https://babeljs.io/docs/usage/babelrc/) yay!) that this file is only used when transpiling src/ to lib/ in the prepublish phase, and should be kept in sync with the babel-loader configuration (`query`) in webpack.config.js?

tbroyer commented on package.json:
```diff
@@ -2,19 +2,20 @@
     "build": "webpack -p",
     "start": "webpack-dev-server -d --inline --hot",
-    "prepublish": "npm run build"
+    "prepublish": "npm run build && BABEL_DISABLE_CACHE=1 babel src --out-dir lib"
```
Should this run `npm run test` instead of `npm run build`?

autra replied:
Or build should be `npm run lint && webpack -p && npm run doc`?

autra replied:
So this conversation came back, as I reverted some change. I actually don't know :-) 

Maybe test should run `lint` and `doclint` only? And then prepublish could run `build`, `test` and babel? But as we run lint and tests from a PR, it's not very important for it to do so (though it gives another layer of security of course). WDYT?

tbroyer replied:
OK, let's use `npm run build` (I kind-of like having everything properly linted and tested before releasing, which is why I proposed using `npm run test`; but then we'd probably want to go the last mile of also checking that the current checkout is "clean", and properly tagged as as the appropriate version, etc. Those might be better in a "release script" than in the `prepublish` script)

tbroyer commented on webpack-babel.config.js:
```diff
@@ -0,0 +1,20 @@
+
+module.exports = {
+  output: {
+    libraryTarget: 'umd',
```
Why UMD? The target is NodeJS here, so CommonJS (from the [`babel-plugin-webpack-loaders` examples](https://github.com/istarkov/minimal-example-for-babel-plugin-webpack-loaders/blob/c901553c80024b4e7fdf961228b0ef55de53f230/webpack.config.js#L4), this “NEEDS TO” be `commonjs2`)

autra replied:
Changing...

autra replied:
Actually answering the `why UMD` question: because it works and because we use UMD in `webpack.config.js`. But let's follow the instructions you linked.

autra replied:
Btw, are we sure the target is nodejs? ppl will `import` or `require` our sources from a browser right? Or maybe I misunderstood something?

tbroyer replied:
Files from `lib/` will be used like any other NPM module; i.e. passed through Webpack, Browserify, Rollup, or equivalent (or in case there's a very good browser-environment emulation –jsdom?– that can do WebGL, then it could be NodeJS proper), which will rewrite the CommonJS imports.

People using iTowns by other means will use the output from Webpack (in `dist/`, as a UMD module, so compatible with RequireJS/AMD **and** direct use in a `<script src=>`, and also incidentally compatible with CommonJS)

autra commented:
@tbroyer updated. WDYT?

tbroyer commented on package.json:
```diff
@@ -2,19 +2,20 @@
     "build": "webpack -p",
     "start": "webpack-dev-server -d --inline --hot",
-    "prepublish": "npm run build"
+    "prepublish": "npm run lint && npm run doc && npm run build && BABEL_DISABLE_CACHE=1 babel src --out-dir lib"
```
Where does this (`npm run doc`) generate the documentation? shouldn't it rather be `npm run doclint` instead (only to ensure the jsdoc is OK)?  
Do we package the doc in the NPM module? (and if so, should we? I'm not quite sure…)

autra replied:
> Where does this (npm run doc) generate the documentation?

I thought it would generate it in `doc`, but actually it does it in `out`, except in `autodeploy.sh` for publishing to gh_pages.

So we don't actually package it. I'm not sure we need to, so I'm going to remove this `npm run doc` from the prepublish phase.

> shouldn't it rather be npm run doclint instead (only to ensure the jsdoc is OK)?

it is already run in the `test` phase (so run for each PR).  If we don't package the doc, I guess we don't need to lint it there.

Thanks for the remark about the doc, I now thing we should just do:
`npm run build && BABEL_DISABLE_CACHE=1 babel src --out-dir lib`. 

WDYT?

tbroyer commented on package.json:
```diff
@@ -2,19 +2,20 @@
   "files": [
     "*.md",
     "dist",
+    "lib/*",
```
Why use `/*` here? other directories are listed as simple entries without `/*` suffix.

autra replied:
Removing it.

tbroyer reviewed (approved):
Good to go, with a couple small comments. Feel free to not address them.

autra commented:
@tbroyer Only the `prepublish` script to finalize :-)

gchoqueux commented on package.json:
```diff
@@ -27,31 +28,30 @@
     "url": "https://github.com/iTowns/itowns2/issues"
   },
   "homepage": "https://itowns.github.io/",
+  "dependencies": {
```
There is a return on the `dependencies`? 
[https://github.com/iTowns/itowns2/pull/241/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L23](https://github.com/iTowns/itowns2/pull/241/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L23)

tbroyer replied:
Yes, that's exactly what this PR is about: consuming the NPM module like any other library, with its own dependencies, rather than (necessarily) as a bundle (and then bundle things yourself using webpack or browserify; and properly deduplicate those dependencies used by your own code and by iTowns)

gchoqueux reviewed (commented):
change commit name 
> type (scope): subject

gchoqueux commented:
Everyone agrees to merge this one?

autra commented:
Rebased 

> Everyone agrees to merge this one?

I don't know, but at least nobody seems to be against it ;-)
